[RES-374] Separate threshold parameters in column pooler

ywcui1990 opened the pull request:
@subutai 

Implements RES-274

* Introduces separate set of threshold parameters for proximal and distal dendrites
Note that activationThreshold exists only for the distal dendrites, as the proximal activation is determined with spatial pooler like rules.

* Introduces separate set of maxNewSynapsesCount and maxSynapsesCount for proximal and distal dendrites. Typically maxSynapsesCount should be higher on proximal dendrites than the distal dendrite as a single proximal zone has to pool over many L4 inputs

* Make sure number of proximal synapses is less than maxNewProximalSynapseCount. Proximal synapse will stop learning if maxNewProximalSynapseCount is achieved.

* Verify that existing scripts still run after the change.

subutai commented on htmresearch/algorithms/column_pooler.py:
```diff
@@ -403,26 +410,89 @@ def _learnProximal(self,
-      cellPermanencesDense[grow] = initialPermanence
-
-      # Update proximalPermanences and proximalConnections.
+      cellNonZeroIndices = list(cellNonZeroIndices)
```
@ywcui1990 Looks like you are reverting some optimization changes here that were recently made by @mrcslws . Can you first merge with master in your branch and then update the PR?

subutai commented on htmresearch/algorithms/column_pooler.py:
```diff
@@ -403,26 +410,89 @@ def _learnProximal(self,
 
 
+
+  def _pickProximalInputsToLearnOn(self, newSynapseCount, activeInputs,
```
This function was removed in recent changes

subutai commented on htmresearch/frameworks/layers/l2_l4_inference.py:
```diff
@@ -544,10 +544,14 @@ def getDefaultL2Params(self, inputSize):
       "initialProximalPermanence": 0.6,
-      "minThreshold": 10,
+      "minThresholdDistal": 10,
+      "minThresholdProximal": 1,
```
Why is the default value 1?  That won't be robust to noise, will it?

subutai commented:
@ywcui1990 I left a few comments. In addition can you make sure tests and the other scripts in projects/l2_pooling continue to run after these changes?   Tests are in tests/algorithms, tests/regions, and tests/layers. You can them from tests using `py.test layers algorithms regions`

ywcui1990 commented:
@subutai  I addressed your comments. I verified that the test passes, except for layers/physical_objects_test.py, which fails before I made the parameter changes for an unrelated reason. I will look at that issue separately.

subutai commented:
👍
